File: src/problem00648/solution648.js
/**
 * 648. Replace Words
 *
 * - Difficulty: Medium
 * - Topics: Array, Hash Table, String, Trie
 *
 * ------------------
 * Author: F. Waskito
 */

/**
 * @param {string[]} dictionary
 * @param {string} sentence
 * @return {string}
 */
const replaceWords = function (dictionary, sentence) {
  let words = sentence.split(" ");
  for (let [i, word] of words.entries()) {
    let shortest_root = "";
    let shortest_root_len = word.length;

    for (let root of dictionary) {
      if (word.startsWith(root) && root.length < shortest_root_len) {
        shortest_root = root;
        shortest_root_len = root.length;
      }
    }

    if (shortest_root) words[i] = shortest_root;
  }

  return words.join(" ");
};

// let dictionary = ["cat", "bat", "rat"];
// let sentence = "the cattle was rattled by the battery";
// expected: "the cat was rat by the bat"

// lst dictionary = ["a", "b", "c"];
// let sentence = "aadsfasf absbs bbab cadsfafs";
// expected: "a a b c"

let dictionary = ["catt", "cat", "bat", "rat"];
let sentence = "the cattle was rattled by the battery";
// expected: "the catt was rat by the bat"

console.log(replaceWords(dictionary, sentence));

File: src/problem00648/solution648.py
"""
648. Replace Words

- Diffulty: Medium
- Topics: Array, Hash Table, String, Trie

------------------
Author: F. Waskito
"""

from typing import List


class Solution:
    def replaceWords(self, dictionary: List[str], sentence: str) -> str:
        words = sentence.split(" ")
        for i, word in enumerate(words):
            shortest_root = ""
            shortest_root_len = len(word)

            for root in dictionary:
                if word.startswith(root) and (len(root) < shortest_root_len):
                    shortest_root = root
                    shortest_root_len = len(root)

            if shortest_root:
                words[i] = shortest_root

        return " ".join(words)


if __name__ == "__main__":
    # dictionary_ = ["cat", "bat", "rat"]
    # sentence_ = "the cattle was rattled by the battery"
    # expected: "the cat was rat by the bat"

    # dictionary_ = ["a", "b", "c"]
    # sentence_ = "aadsfasf absbs bbab cadsfafs"
    # expected: "a a b c"

    dictionary_ = ["catt", "cat", "bat", "rat"]
    sentence_ = "the cattle was rattled by the battery"
    # expected: "the catt was rat by the bat"

    print(Solution().replaceWords(dictionary_, sentence_))
